<a href="https://cognitiveclass.ai"><img src = "https://ibm.box.com/shared/static/9gegpsmnsoo25ikkbl4qzlvlyjbgxs5x.png" width = 400> </a>

# Opening Restaurant in London

## 1. Business Problem

According to www.food.gov.uk, there are more than 14,000 restaurants in London and about 9 million people. That is why opening a new restaurant there can be an extremely challenging task. Choosing a restaurant type and a good spot, an entrepreneur usually carelessly relies on common sense and domain knowledge. Needless to say that too often an inconsiderate decision leads to a poor income and inevitable bankruptcy. According to several surveys, up to 40% of such start-ups fail in the very first year. Let's suppose, an investor has enough time and money, as well as a passion to open the best eating spot in London. What type of restaurant would it be? What would be the best place for it? Is there a better way to answer these questions rather than guessing?  
What if there is a way to cluster city neighborhoods, based on their near-by restaurant similarity? What if we can visualize these clusters on a map? What if we might find what type of restaurant is the most and least popular in each location? Equipped with that knowledge, we might be able to make a smart choice from a huge number of restaurant types and available places.  
Let us allow machine learning to get the job done. Using reliable venue data, it can investigate the city neighborhoods, and show us unseen dependencies. Dependencies that we are not aware of.


**Target audience:** investors, entrepreneurs, and chefs interested in opening a restaurant in London, who may need a piece of objective advice of what type of restaurant would be more successful and where exactly it should be opened.

## 2. Data

**Step 1.** Using a table on https://en.wikipedia.org/wiki/List_of_areas_of_London, collect information about London boroughs and locations, excluding records whose "Post Town" is not London.  
**Step 2.** Use the Geopy and Folium library to get the coordinates of every locations and map geospatial data on a London map.  
**Step 3.** Using Foursquare API, collect the top 100 restaurants and their categories for each location within a radius 500 meters.  
**Step 4.** Group collected restaurants by location and by taking the mean of the frequency of occurrence of each type, preparing them for clustering.   
**Step 5.** Cluster restaurants by k-means algorithm and analyze the top 10 most common restaurants in each cluster.    
**Step 6.** Visualize clusters on the map, thus showing the best locations for opening the chosen restaurant.

## 3. Methodology

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [101]:
import time # for time delay while working with API

import requests # library to handle requests

import bs4 # library to parse webpages

import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Convert an address into latitude and longitude values
from geopy.geocoders import Nominatim
import geopy.geocoders

import json # library to handle JSON files

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# k-means from clustering stage
from sklearn.cluster import KMeans

# Map rendering library
import folium

# regular expressions
import re

### 3.1. Collecting London Neighborhoods

Let's create a webscrapping script to collect London neighborhoods information from the table on https://en.wikipedia.org/wiki/List_of_areas_of_London with following columns: Post_town, Borough, and Location.

In [102]:
# Download the webpage
url = 'https://en.wikipedia.org/wiki/List_of_areas_of_London'
res = requests.get(url)
res.raise_for_status()

In [103]:
# Create an beautifulSoup object
london_soup = bs4.BeautifulSoup(res.text)

In [104]:
# Selecting all elements inside the corresponding tags
elements = london_soup.select('div table tbody tr td')

In [105]:
# Take a look on raw data
for i in range(2, len(elements), 6):
    print('{0} | {1} | {2} | {3}'.format(str(i//6+1), elements[i].getText(), elements[i+1].getText(), elements[i+2].getText(),
                                                    elements[i+3].getText()))
    if elements[i].getText() == 'Yiewsley': # the last location on the table
        break

1 | Abbey Wood | Greenwich[1] | LONDON
2 | Acton | Ealing, Hammersmith and Fulham[2] | LONDON
3 | Addington | Croydon[2] | CROYDON
4 | Addiscombe | Croydon[2] | CROYDON
5 | Albany Park | Bexley | BEXLEY, SIDCUP
6 | Aldborough Hatch | Redbridge[3] | ILFORD
7 | Aldgate | City[4] | LONDON
8 | Aldwych | Westminster[4] | LONDON
9 | Alperton | Brent[5] | WEMBLEY
10 | Anerley | Bromley[5] | LONDON
11 | Angel | Islington[2] | LONDON
12 | Aperfield | Bromley[5] | WESTERHAM
13 | Archway | Islington[6] | LONDON
14 | Ardleigh Green | Havering[6] | HORNCHURCH
15 | Arkley | Barnet[6] | BARNET, LONDON
16 | Arnos Grove | Enfield[6] | LONDON
17 | Balham | Wandsworth[7] | LONDON
18 | Bankside | Southwark[8] | LONDON
19 | Barbican | City[8] | LONDON
20 | Barking | Barking and Dagenham[8] | BARKING
21 | Barkingside | Redbridge[9] | ILFORD
22 | Barnehurst | Bexley[9] | BEXLEYHEATH
23 | Barnes | Richmond upon Thames[9] | LONDON
24 | Barnes Cray | Bexley[10] | DARTFORD
25 | Barnet Gate | Barnet | LONDON, BAR

In [106]:
yiewsley_index = (533-1)*6 + 2
elements[yiewsley_index].get_text()

'Yiewsley'

At the previous step we collected 533 rows with data. The last location in the table is 'Yiewsley' and its index in the _elements_ list is 3194. Let's transform raw data into a list of lists, considering the restriction to ignore location with a _Postal Town_ that is not 'LONDON'. Also we will add two zeros in each row as a initial geographical coordinates.

In [107]:
# Creating a new list of rows
lst = []
for i in range(2, 3195, 6):
    location, borough, postal_town = elements[i].getText(), elements[i+1].getText(), elements[i+2].getText()
    if postal_town != 'LONDON':
        continue
    lst.append([location, borough, postal_town, 0, 0])
lst[25:34]

[['Bloomsbury', 'Camden[23]', 'LONDON', 0, 0],
 ['Bounds Green', 'Haringey[25]', 'LONDON', 0, 0],
 ['Bow', 'Tower Hamlets[25]', 'LONDON', 0, 0],
 ['Bowes Park', 'Haringey[26]', 'LONDON', 0, 0],
 ['Brent Cross', 'Barnet', 'LONDON', 0, 0],
 ['Brent Park', 'Brent', 'LONDON', 0, 0],
 ['Brixton', 'Lambeth[28]', 'LONDON', 0, 0],
 ['Brockley', 'Lewisham[28]', 'LONDON', 0, 0],
 ['Bromley (also Bromley-by-Bow)', 'Tower Hamlets[30]', 'LONDON', 0, 0]]

As we can see there is some garbage in or data, for example in the last row in the previous output: ['Bromley (also Bromley-by-Bow)', 'Tower Hamlets[30]', 'LONDON'].  
Let's clean our data by deleting text in brackets using regular expressions.

In [108]:
for i in range(len(lst)):
    loc, bor = lst[i][0], lst[i][1]
    if loc.endswith(')') or loc.endswith(']'):
        lst[i][0] = re.sub('(\s?\(.*?\)$)|(\s?\[.*?\]$)', '', loc)
    if bor.endswith(')') or bor.endswith(']'):
        lst[i][1] = re.sub('(\s?\(.*?\)$)|(\s?\[.*?\]$)', '', bor)
lst[25:34]

[['Bloomsbury', 'Camden', 'LONDON', 0, 0],
 ['Bounds Green', 'Haringey', 'LONDON', 0, 0],
 ['Bow', 'Tower Hamlets', 'LONDON', 0, 0],
 ['Bowes Park', 'Haringey', 'LONDON', 0, 0],
 ['Brent Cross', 'Barnet', 'LONDON', 0, 0],
 ['Brent Park', 'Brent', 'LONDON', 0, 0],
 ['Brixton', 'Lambeth', 'LONDON', 0, 0],
 ['Brockley', 'Lewisham', 'LONDON', 0, 0],
 ['Bromley', 'Tower Hamlets', 'LONDON', 0, 0]]

So our dataset is clear enough and ready to be transformed into a pandas dataframe. Fine! By the way, how many locations do we have?

In [109]:
print('Now we have {} rows of relevant data.'.format(len(lst)))

Now we have 299 rows of relevant data.


Let's transform them.

In [110]:
london_df = pd.DataFrame(lst, columns=['Location', 'Borough', 'PostalTown', 'Latitude', 'Longitude'])
london_df.head()

Location                         Borough PostalTown  Latitude  Longitude
0  Abbey Wood                       Greenwich     LONDON         0          0
1       Acton  Ealing, Hammersmith and Fulham     LONDON         0          0
2     Aldgate                            City     LONDON         0          0
3     Aldwych                     Westminster     LONDON         0          0
4     Anerley                         Bromley     LONDON         0          0

Confirm the size:

In [111]:
london_df.shape

(299, 5)

### 3.2. Adding Coordinates

In order to utilize the Foursquare location data, we need to get latitude and longitude coordinates for each neighborhood in the dataframe.  
We will use the geopy library for that purpose. Let's try with the first address that is Abbey Wood, Greenwich, London.

In [112]:
# Getting the address string
address = ', '.join(list(london_df.iloc[0, :3]))
address

'Abbey Wood, Greenwich, LONDON'

In [113]:
# Using geopy
geolocator = Nominatim(user_agent='opening_restaurant_london')
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of {0} are {1}, {2}.'.format(address, latitude, longitude))

The geograpical coordinate of Abbey Wood, Greenwich, LONDON are 51.487621, 0.1140504.


In [114]:
# Make changes to the dataframe
london_df.iloc[0,3] = latitude
london_df.iloc[0,4] = longitude
london_df.head()

Location                         Borough PostalTown   Latitude  Longitude
0  Abbey Wood                       Greenwich     LONDON  51.487621    0.11405
1       Acton  Ealing, Hammersmith and Fulham     LONDON   0.000000    0.00000
2     Aldgate                            City     LONDON   0.000000    0.00000
3     Aldwych                     Westminster     LONDON   0.000000    0.00000
4     Anerley                         Bromley     LONDON   0.000000    0.00000

Well done! Now we are ready to apply a for loop to go through all addresses in the dataframe and get the corresponding coordinates.  
**Disclaimer:** due to various API restrictions, the following script takes almost 13 minutes to complete the task. In average it successfully collects 98% coordinates. Saving your time I collected the coordinates and transformed them into the __london_coordinates.csv__ file that you can find in the same GitHub repository.

In [115]:
# Uncomment if you are ready to wait 12-15 minutes.
#for i in range(len(london_df)):
    #time.sleep(2.5)
    #address = ', '.join(list(london_df.iloc[i, :3]))
    #geolocator = Nominatim(user_agent='opening_restaurant_london')
    #location = geolocator.geocode(address)
    #if location == None:
        #continue
    #latitude = location.latitude
    #longitude = location.longitude
    #london_df.iloc[i,3] = latitude
    #london_df.iloc[i,4] = longitude
#london_df

Save the dataframe to the csv format.

In [116]:
#london_df.to_csv('.\london_coordinates.csv', index=False)

And try read it.

In [117]:
lon_df = pd.read_csv('.\london_coordinates.csv')
lon_df.head()

Location                         Borough PostalTown   Latitude  Longitude
0  Abbey Wood                       Greenwich     LONDON  51.487621   0.114050
1       Acton  Ealing, Hammersmith and Fulham     LONDON   0.000000   0.000000
2     Aldgate                            City     LONDON  51.514248  -0.075719
3     Aldwych                     Westminster     LONDON  51.512940  -0.118101
4     Anerley                         Bromley     LONDON  51.412848  -0.065301

The next step is to drop rows that still contain 0 as a latitude or longitude.

In [118]:
# Check initial shape
lon_df.shape

(299, 5)

In [119]:
# Substitute all zeros by NAN
lon_df = lon_df.replace(0, np.nan)

# Drop all rows containing NAN
lon_df.dropna(subset=['Latitude', 'Longitude'], axis=0, inplace=True)
lon_df.reset_index(drop=True, inplace=True)
print('Now the London dataframe has {0} data rows.'.format(lon_df.shape[0]))

Now the London dataframe has 290 data rows.


Check if there are not unique location names.

In [120]:
len(lon_df['Location'].unique())

288

In [121]:
# Printing these locations
for i in range(len(lon_df)):
    loc = lon_df.iloc[i,0]
    for j in range(i+1, len(lon_df)):
        if lon_df.iloc[j,0] == loc:
            print(j, loc)

54 Church End
103 Grove Park


For illustration purposes, let's simplify things and drop the doubled locations.  
_(Actually, we are not allowed to do that because a pair of "not unique" location and its borough is still unique. There are only 2 locations from 290 what is not a big deal.)_

In [122]:
lon_df.drop_duplicates(subset='Location', keep='first', inplace=True)
if lon_df['Location'].unique().shape[0] == lon_df.shape[0]:
    print('Duplicates were removed successfully.')

Duplicates were removed successfully.


Confirm the new size.

In [123]:
lon_df.shape

(288, 5)

So 288 London neighborhoods are ready to be shown on a map.  
We will use the folium library for this purpose.

In [124]:
# Get the London "central" point
london_address = 'London, England'
geolocator = Nominatim(user_agent='opening_restaurant_london')
location = geolocator.geocode(london_address)
london_lat = location.latitude
london_lon = location.longitude
print('The geograpical coordinate of {0} are {1}, {2}.'.format(london_address, london_lat, london_lon))

The geograpical coordinate of London, England are 51.5073219, -0.1276474.


In [125]:
# create map of London using starting point coordinates
london_map = folium.Map(location=[london_lat, london_lon], zoom_start=11)

# add markers to map
for lat, lng, bor, loc in zip(lon_df['Latitude'], lon_df['Longitude'], lon_df['Borough'], lon_df['Location']):
    label = '{}, {}'.format(loc, bor)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        ).add_to(london_map)
    
london_map

## 4. Exploring London Restaurants

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

In [126]:
CLIENT_ID = '5OQ4ZCIYRPY1CSLZRYUTLY2ZWUR4ZOSH2NO1MUW1C3FB0MR5'
CLIENT_SECRET = 'DRQX5Z4UJIUOGZHSOJL104M5ZBN2M2V1XL1XVGOWE5CJK4TF'
VERSION = '20190131'

### 4.1. Collecting Restaurants

Let's explore the first neighborhood in our dataframe.

In [127]:
lon_df.loc[16, 'Location']

'Bellingham'

Get the neighborhood's latitude and longitude values.

In [128]:
loc_latitude = lon_df.loc[16, 'Latitude'] # neighborhood latitude value
loc_longitude = lon_df.loc[16, 'Longitude'] # neighborhood longitude value

loc_name = lon_df.loc[16, 'Location'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(loc_name, 
                                                               loc_latitude, 
                                                               loc_longitude))

Latitude and longitude values of Bellingham are 51.4329965, -0.019337599999999996.


Now, let's get the top 100 venues that are in Marble Hill within a radius of 500 meters.

In [129]:
# type your answer here
radius = 500
LIMIT = 100
url = 'https://api.foursquare.com/v2/venues/explore?client_id={0}&client_secret={1}&ll={2},{3}&v={4}&radius={5}&limit={6}&query=restaurant'.format(CLIENT_ID, CLIENT_SECRET, loc_latitude, loc_longitude, VERSION, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?client_id=5OQ4ZCIYRPY1CSLZRYUTLY2ZWUR4ZOSH2NO1MUW1C3FB0MR5&client_secret=DRQX5Z4UJIUOGZHSOJL104M5ZBN2M2V1XL1XVGOWE5CJK4TF&ll=51.4329965,-0.019337599999999996&v=20190131&radius=500&limit=100&query=restaurant'

In [130]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5c53e977351e3d1df78fb53f'},
 'response': {'headerLocation': 'London',
  'headerFullLocation': 'London',
  'headerLocationGranularity': 'city',
  'query': 'restaurant',
  'totalResults': 4,
  'suggestedBounds': {'ne': {'lat': 51.437496504500004,
    'lng': -0.012132931194960173},
   'sw': {'lat': 51.4284964955, 'lng': -0.02654226880503982}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4b9ce8faf964a520da8136e3',
       'name': 'Turkuaz',
       'location': {'address': '163 Bromley Rd',
        'lat': 51.4353203897812,
        'lng': -0.017901890774046518,
        'labeledLatLngs': [{'label': 'display',
          'lat': 51.4353203897812,
          'lng': -0.017901890774046518}],
        'distance': 277,
        'postalCode': 'SE6 9NZ

In [131]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [132]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name           categories        lat       lng
0              Turkuaz   Turkish Restaurant  51.435320 -0.017902
1          Papa John’s          Pizza Place  51.433615 -0.017601
2  Bellingham Fish Bar    Fish & Chips Shop  51.432443 -0.020708
3             Morley's  Fried Chicken Joint  51.433441 -0.012791

In [133]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

4 venues were returned by Foursquare.


Let's create a function to repeat the same process to all the neighborhoods in London.

In [134]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={0}&client_secret={1}&v={2}&ll={3},{4}&radius={5}&limit={6}&query=restaurant'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            #v['venue']['location']['lat'], 
            #v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Location', 
                  'Latitude', 
                  'Longitude', 
                  'Venue', 
                  #'Venue Latitude', 
                  #'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now run the above function on each neighborhood and create a new dataframe called *london_venues*.

In [135]:
london_venues = getNearbyVenues(names=lon_df['Location'],
                                   latitudes=lon_df['Latitude'],
                                   longitudes=lon_df['Longitude']
                                  )

Let's check the size of the resulting dataframe.

In [136]:
print(london_venues.shape)
london_venues.head()

(6273, 5)


Location   Latitude  Longitude                      Venue  \
0  Abbey Wood  51.487621   0.114050  The Crafty Cafe by Sharon   
1  Abbey Wood  51.487621   0.114050           Frank's Fish Bar   
2     Aldgate  51.514248  -0.075719                  Benk + Bo   
3     Aldgate  51.514248  -0.075719                       Bife   
4     Aldgate  51.514248  -0.075719       The Japanese Canteen   

           Venue Category  
0                    Café  
1       Fish & Chips Shop  
2                  Bakery  
3  Argentinian Restaurant  
4     Japanese Restaurant

Let's check how many restaurants were returned for each neighborhood.

In [153]:
london_venues[['Location', 'Venue']].groupby('Location').count()

Venue
Location                        
Abbey Wood                     2
Aldgate                      100
Aldwych                      100
Anerley                        3
Angel                         65
Archway                       25
Arnos Grove                    7
Balham                        37
Bankside                      60
Barbican                      51
Barnes                         7
Barnsbury                     12
Battersea                     15
Bayswater                     85
Bedford Park                   7
Belgravia                     28
Bellingham                     4
Belsize Park                  28
Bermondsey                    10
Bethnal Green                 29
Blackfriars                   47
Blackheath                    24
Blackheath Royal Standard      4
Blackwall                      6
Bloomsbury                    42
Bounds Green                   6
Bow                            4
Bowes Park                     7
Brent Cross                   16
Brent Park                     4
Brixton                       10
Brockley                      13
Bromley                        4
Brompton                      12
Brondesbury                   20
Brunswick Park                 3
Burroughs, The                12
Camberwell                    36
Cambridge Heath               21
Camden Town                   61
Canary Wharf                  95
Canning Town                   7
Canonbury                      9
Castelnau                      6
Catford                       14
Chalk Farm                    32
Charing Cross                100
Charlton                       7
Chelsea                       39
Childs Hill                    4
Chinatown                    100
Chingford                      8
Church End                    16
Clapham                       11
Clerkenwell                  100
Colindale                      4
Colliers Wood                 14
Colney Hatch                   7
Covent Garden                100
Cricklewood                   17
Crofton Park                   9
Crouch End                    39
Crystal Palace                 4
Cubitt Town                    6
Custom House                   8
Dalston                        6
Dartford                       2
De Beauvoir Town              13
Denmark Hill                   9
Deptford                      17
Dollis Hill                    7
Dulwich                        2
Ealing                        54
Earls Court                   44
Earlsfield                     8
East Dulwich                  10
East Finchley                 14
East Ham                       7
East Sheen                    10
Edmonton                       7
Elephant and Castle           30
Eltham                         5
Farringdon                    58
Finchley                       3
Finsbury                      61
Finsbury Park                 38
Fitzrovia                    100
Forest Gate                   10
Forest Hill                   11
Fortis Green                   7
Friern Barnet                  5
Frognal                       29
Fulham                        36
Gipsy Hill                     4
Golders Green                 29
Gospel Oak                     5
Grahame Park                   2
Grange Park                    3
Greenwich                     38
Grove Park                     5
Gunnersbury                   14
Hackney                       12
Hackney Central               44
Hackney Marshes                1
Hackney Wick                  18
Haggerston                    19
Hammersmith                   61
Hampstead                     32
Hanwell                        6
Harlesden                      4
Harringay                     23
Hendon                        16
Herne Hill                    13
Highams Park                  10
Highbury                      46
Highgate                       5
Hither Green                   8
Holborn                       62
Holland Park                  15
Holloway                      31
H

And check if Foursquare API did not return restaurants for some locations.

In [166]:
x = london_venues[['Location', 'Venue']].groupby('Location').count().shape[0]
y = lon_df.shape[0]
empty_locations = []
if x != y:
    print('Missing data for {0} locations:'.format(y-x))
    # And print them
    for i in range(lon_df.shape[0]):
        loc = lon_df.iloc[i,0]
        k = 0
        for j in range(london_venues.shape[0]):
            if loc == london_venues.iloc[j,0]:
                k += 1
        if k == 0:
            print(i,loc)
            empty_locations.append(loc)

Missing data for 5 locations:
51 Chinbrook
62 Crossness
111 Hampstead Garden Suburb
163 Mill Hill
247 Totteridge


Let's find out how many unique categories can be curated from all the returned restaurants.

In [138]:
print('There are {0} uniques categories.'.format(len(london_venues['Venue Category'].unique())))

There are 126 uniques categories.


### 4.2. Exploring Restaurants

To begin analisys we need to transform collected information using the one-hot encoding method.

In [139]:
# one hot encoding
london_onehot = pd.get_dummies(london_venues[['Venue Category']], prefix="", prefix_sep="")

# add location column back to dataframe
london_onehot['Location'] = london_venues['Location'] 

# move location column to the first column
fixed_columns = [london_onehot.columns[-1]] + list(london_onehot.columns[:-1])
london_onehot = london_onehot[fixed_columns]

london_onehot.head()

Location  Afghan Restaurant  African Restaurant  American Restaurant  \
0  Abbey Wood                  0                   0                    0   
1  Abbey Wood                  0                   0                    0   
2     Aldgate                  0                   0                    0   
3     Aldgate                  0                   0                    0   
4     Aldgate                  0                   0                    0   

   Arepa Restaurant  Argentinian Restaurant  Asian Restaurant  \
0                 0                       0                 0   
1                 0                       0                 0   
2                 0                       0                 0   
3                 0                       1                 0   
4                 0                       0                 0   

   Australian Restaurant  Austrian Restaurant  BBQ Joint  Bagel Shop  Bakery  \
0                      0                    0          0           0       0   
1                      0                    0          0           0       0   
2                      0                    0          0           0       1   
3                      0                    0          0           0       0   
4                      0                    0          0           0       0   

   Belgian Restaurant  Bistro  Brazilian Restaurant  Breakfast Spot  \
0                   0       0                     0               0   
1                   0       0                     0               0   
2                   0       0                     0               0   
3                   0       0                     0               0   
4                   0       0                     0               0   

   Bulgarian Restaurant  Burger Joint  Burrito Place  Cafeteria  Café  \
0                     0             0              0          0     1   
1                     0             0              0          0     0   
2                     0             0              0          0     0   
3                     0             0              0          0     0   
4                     0             0              0          0     0   

   Cajun / Creole Restaurant  Cantonese Restaurant  Caribbean Restaurant  \
0                          0                     0                     0   
1                          0                     0                     0   
2                          0                     0                     0   
3                          0                     0                     0   
4                          0                     0                     0   

   Caucasian Restaurant  Chaat Place  Chinese Restaurant  Churrascaria  \
0                     0            0                   0             0   
1                     0            0                   0             0   
2                     0            0                   0             0   
3                     0            0                   0             0   
4                     0            0                   0             0   

   Cigkofte Place  Comfort Food Restaurant  Creperie  Cuban Restaurant  \
0               0                        0         0                 0   
1               0                        0         0                 0   
2               0                        0         0                 0   
3               0                        0         0                 0   
4               0                        0         0                 0   

   Currywurst Joint  Deli / Bodega  Dim Sum Restaurant  Diner  \
0                 0              0                   0      0   
1                 0              0                   0      0   
2                 0              0                   0      0   
3                 0              0                   0      0   
4                 0              0                   0      0   

   Doner Restaurant  Donut Shop  Dumpling Restaurant  \
0         

And let's examine the new dataframe size.

In [140]:
london_onehot.shape

(6273, 127)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category preparing the dataframe for clustering.

In [141]:
london_grouped = london_onehot.groupby('Location').mean().reset_index()
london_grouped

Location  Afghan Restaurant  African Restaurant  \
0                   Abbey Wood           0.000000            0.000000   
1                      Aldgate           0.000000            0.000000   
2                      Aldwych           0.000000            0.000000   
3                      Anerley           0.000000            0.000000   
4                        Angel           0.015385            0.000000   
5                      Archway           0.000000            0.000000   
6                  Arnos Grove           0.000000            0.000000   
7                       Balham           0.000000            0.000000   
8                     Bankside           0.000000            0.000000   
9                     Barbican           0.000000            0.000000   
10                      Barnes           0.000000            0.000000   
11                   Barnsbury           0.000000            0.083333   
12                   Battersea           0.000000            0.000000   
13                   Bayswater           0.000000            0.011765   
14                Bedford Park           0.000000            0.000000   
15                   Belgravia           0.000000            0.000000   
16                  Bellingham           0.000000            0.000000   
17                Belsize Park           0.000000            0.000000   
18                  Bermondsey           0.000000            0.000000   
19               Bethnal Green           0.000000            0.000000   
20                 Blackfriars           0.000000            0.000000   
21                  Blackheath           0.000000            0.000000   
22   Blackheath Royal Standard           0.000000            0.000000   
23                   Blackwall           0.000000            0.000000   
24                  Bloomsbury           0.000000            0.000000   
25                Bounds Green           0.000000            0.000000   
26                         Bow           0.000000            0.000000   
27                  Bowes Park           0.000000            0.000000   
28                 Brent Cross           0.000000            0.000000   
29                  Brent Park           0.000000            0.000000   
30                     Brixton           0.000000            0.000000   
31                    Brockley           0.000000            0.000000   
32                     Bromley           0.000000            0.000000   
33                    Brompton           0.000000            0.000000   
34                 Brondesbury           0.000000            0.000000   
35              Brunswick Park           0.000000            0.000000   
36              Burroughs, The           0.000000            0.000000   
37                  Camberwell           0.000000            0.000000   
38             Cambridge Heath           0.000000            0.000000   
39                 Camden Town           0.000000            0.016393   
40                Canary Wharf           0.000000            0.000000   
41                Canning Town           0.000000            0.000000   
42                   Canonbury           0.000000            0.000000   
43                   Castelnau           0.000000            0.000000   
44                     Catford           0.000000            0.000000   
45                  Chalk Farm           0.000000            0.000000   
46               Charing Cross           0.000000            0.020000   
47                    Charlton           0.000000            0.000000   
48                     Chelsea           0.000000            0.000000   
49                 Childs Hill           0.000000            0.000000   
50                   Chinatown           0.000000            0.010000   
51                   Chingford           0.000000            0.000000   
52                  Church End           0.000000            0.000000   
53                     Clapham           0.000000            0.000000   
54     

Let's confirm the new size.

In [142]:
london_grouped.shape

(283, 127)

Let's investigate each neighborhood along with the top 5 most common venues.

In [143]:
# Function to sort the venues in descending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [144]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Location']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Location'] = london_grouped['Location']

for ind in np.arange(london_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(london_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Location          1st Most Common Venue  \
0                   Abbey Wood                           Café   
1                      Aldgate                           Café   
2                      Aldwych                     Restaurant   
3                      Anerley                    Pizza Place   
4                        Angel                           Café   
5                      Archway                           Café   
6                  Arnos Grove                     Steakhouse   
7                       Balham                           Café   
8                     Bankside             Italian Restaurant   
9                     Barbican             Italian Restaurant   
10                      Barnes                     Restaurant   
11                   Barnsbury                           Café   
12                   Battersea                           Café   
13                   Bayswater                           Café   
14                Bedford Park                           Café   
15                   Belgravia                           Café   
16                  Bellingham             Turkish Restaurant   
17                Belsize Park                           Café   
18                  Bermondsey                           Café   
19               Bethnal Green                           Café   
20                 Blackfriars                 Sandwich Place   
21                  Blackheath              Indian Restaurant   
22   Blackheath Royal Standard                  Deli / Bodega   
23                   Blackwall             Italian Restaurant   
24                  Bloomsbury                           Café   
25                Bounds Green              Indian Restaurant   
26                         Bow                   Burger Joint   
27                  Bowes Park              Indian Restaurant   
28                 Brent Cross                           Café   
29                  Brent Park              Indian Restaurant   
30                     Brixton              Indian Restaurant   
31                    Brockley                           Café   
32                     Bromley                   Burger Joint   
33                    Brompton                           Café   
34                 Brondesbury              Indian Restaurant   
35              Brunswick Park                          Diner   
36              Burroughs, The               Sushi Restaurant   
37                  Camberwell                           Café   
38             Cambridge Heath                     Restaurant   
39                 Camden Town  Vegetarian / Vegan Restaurant   
40                Canary Wharf                 Sandwich Place   
41                Canning Town             English Restaurant   
42                   Canonbury                           Café   
43                   Castelnau                           Café   
44                     Catford                           Café   
45                  Chalk Farm                           Café   
46               Charing Cross                           Café   
47                    Charlton              Indian Restaurant   
48                     Chelsea            Japanese Restaurant   
49                 Childs Hill             Italian Restaurant   
50                   Chinatown                   Burger Joint   
51                   Chingford             Italian Restaurant   
52                  Church End                           Café   
53                     Clapham           Fast Food Restaurant   
54                 Clerkenwell                           Café   
55                   Colindale                    Pizza Place   
56               Colliers Wood              Indian Restaurant   
57                Colney Hatch              Indian Restaurant   
58               Covent Garden              French Restaurant   
59                 Cricklewood           Fast Food Restaurant   
60                Crofton Park                           

### 4.3. Clustering Restaurants

Run *k*-means to cluster the neighborhood into 5 clusters.

In [145]:
# set number of clusters
kclusters = 5

london_grouped_clustering = london_grouped.drop('Location', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=4).fit(london_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 1, 1, 1, 1, 4, 1, 1, 1, 1])

Let's create a new dataframe that includes the clusters as well as the top 10 venues for each neighborhood.  
Do not forget that some location didn't get any data from Foursquare API, and we put them to the list.  
Therfore we are forced to exclude them from the resulting dataset.

In [167]:
london_merged = lon_df

# Substitute all empty locations by NAN
for loc in empty_locations:
    london_merged = london_merged.replace(loc, np.nan)

# then drop all rows containing NAN
london_merged.dropna(subset=['Location'], axis=0, inplace=True)
london_merged.reset_index(drop=True, inplace=True)
print('Now the cluster dataframe has {0} data rows.'.format(london_merged.shape[0]))

# add clustering labels
london_merged['Cluster Labels'] = kmeans.labels_

# merge london_grouped with lon_df to add latitude/longitude for each neighborhood
london_merged = london_merged.join(neighborhoods_venues_sorted.set_index('Location'), on='Location')

london_merged.head()

Now the London dataframe has 283 data rows.


Location      Borough PostalTown   Latitude  Longitude  Cluster Labels  \
0  Abbey Wood    Greenwich     LONDON  51.487621   0.114050               2   
1     Aldgate         City     LONDON  51.514248  -0.075719               1   
2     Aldwych  Westminster     LONDON  51.512940  -0.118101               1   
3     Anerley      Bromley     LONDON  51.412848  -0.065301               1   
4       Angel    Islington     LONDON  51.531946  -0.106106               1   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                  Café     Fish & Chips Shop    Yoshoku Restaurant   
1                  Café     Indian Restaurant      Sushi Restaurant   
2            Restaurant        Sandwich Place     French Restaurant   
3           Pizza Place        Sandwich Place    Chinese Restaurant   
4                  Café    Italian Restaurant            Restaurant   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0      Currywurst Joint    Dim Sum Restaurant                 Diner   
1        Sandwich Place           Salad Place            Restaurant   
2    Italian Restaurant                  Café      Sushi Restaurant   
3    Yoshoku Restaurant   Filipino Restaurant    Dim Sum Restaurant   
4        Sandwich Place      Sushi Restaurant         Burrito Place   

           7th Most Common Venue  8th Most Common Venue 9th Most Common Venue  \
0               Doner Restaurant             Donut Shop   Dumpling Restaurant   
1                Thai Restaurant     Italian Restaurant   Japanese Restaurant   
2              Indian Restaurant           Burger Joint            Steakhouse   
3                          Diner       Doner Restaurant            Donut Shop   
4  Vegetarian / Vegan Restaurant  Vietnamese Restaurant          Burger Joint   

        10th Most Common Venue  
0  Eastern European Restaurant  
1    Middle Eastern Restaurant  
2          American Restaurant  
3          Dumpling Restaurant  
4            French Restaurant

## 5. Results

And now we are ready to conclude our report.

### 5.1. Examine Clusters

Let's examine each cluster and the discriminating restaurant categories that distinguish a cluster.

#### Cluster 1

In [173]:
london_merged.loc[london_merged['Cluster Labels'] == 0, london_merged.columns[[1] + list(range(5, london_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22               Greenwich               0         Deli / Bodega   
26           Tower Hamlets               0          Burger Joint   
32           Tower Hamlets               0          Burger Joint   
53     Lambeth, Wandsworth               0  Fast Food Restaurant   
59   Barnet, Brent, Camden               0  Fast Food Restaurant   
99                Hounslow               0  Fast Food Restaurant   
133          Brent, Camden               0                  Café   
150                Hackney               0                  Café   
189            Westminster               0     Indian Restaurant   
198              Southwark               0                  Café   
201               Haringey               0           Pizza Place   
231                Lambeth               0                  Café   
236    Lewisham, Southwark               0                  Café   
237      City, Westminster               0            Restaurant   
261                 Ealing               0  Fast Food Restaurant   
278               Haringey               0  Fast Food Restaurant   

    2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
22     Chinese Restaurant                       Café  Fast Food Restaurant   
26   Fast Food Restaurant                       Café    Yoshoku Restaurant   
32   Fast Food Restaurant                       Café    Yoshoku Restaurant   
53      Indian Restaurant                  Gastropub    Chinese Restaurant   
59         Breakfast Spot                       Café     Afghan Restaurant   
99         Breakfast Spot                       Café    Chinese Restaurant   
133           Pizza Place         Italian Restaurant   Fried Chicken Joint   
150              Creperie                  Gastropub   Dumpling Restaurant   
189           Pizza Place         Italian Restaurant        Sandwich Place   
198        Breakfast Spot                 Restaurant    Chinese Restaurant   
201        Breakfast Spot  South American Restaurant   Fried Chicken Joint   
231    Chinese Restaurant                Pizza Place  Fast Food Restaurant   
236            Food Truck         Yoshoku Restaurant     Fish & Chips Shop   
237                  Café             Sandwich Place     French Restaurant   
261     Indian Restaurant                      Diner      Doner Restaurant   
278                  Café         Chinese Restaurant            Restaurant   

    5th Most Common Venue  6th Most Common Venue      7th Most Common Venue  \
22     Yoshoku Restaurant                  Diner           Doner Restaurant   
26          Deli / Bodega                  Diner           Doner Restaurant   
32          Deli / Bodega                  Diner           Doner Restaurant   
53   Caucasian Restaurant      Fish & Chips Shop             Breakfast Spot   
59     Italian Restaurant   Caribbean Restaurant           Asian Restaurant   
99     Yoshoku Restaurant     English Restaurant        Filipino Restaurant   
133         Deli / Bodega  Portuguese Restaurant           Doner Restaurant   
150          Burger Joint            Pizza Place  Latin American Restaurant   
189                  Café     Turkish Restaurant         Chinese Restaurant   
198        Sandwich Place    Szechuan Restaurant         Yoshoku Restaurant   
201    Yoshoku Restaurant    Empanada Restaurant         Falafel Restaurant   
231                Bakery       Sushi Restaurant               Burger Joint   
236    Dim Sum Restaurant                  Diner           Doner Restaurant   
237    Italian Restaurant       Sushi Restaurant        Japanese Restaurant   
261                  Café                 Bistro         Turkish Restaurant   
278        Sandwich Place     Turkish Restaurant   Mediterranean Restaurant   

    8th Most Common Venue          9th Most Common Venue  \
22             Donut Shop            Dumpling Restaurant   
26             Donut Shop            Dumpling Restaurant   
32         

In [180]:
cluster_1 = london_merged.loc[london_merged['Cluster Labels'] == 0, london_merged.columns[[1] + list(range(5, london_merged.shape[1]))]]
cluster_1.describe(include='all')

Borough  Cluster Labels 1st Most Common Venue  \
count              16            16.0                    16   
unique             14             NaN                     7   
top     Tower Hamlets             NaN                  Café   
freq                2             NaN                     5   
mean              NaN             0.0                   NaN   
std               NaN             0.0                   NaN   
min               NaN             0.0                   NaN   
25%               NaN             0.0                   NaN   
50%               NaN             0.0                   NaN   
75%               NaN             0.0                   NaN   
max               NaN             0.0                   NaN   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
count                     16                    16                    16   
unique                     8                    10                    11   
top           Breakfast Spot                  Café    Chinese Restaurant   
freq                       4                     5                     3   
mean                     NaN                   NaN                   NaN   
std                      NaN                   NaN                   NaN   
min                      NaN                   NaN                   NaN   
25%                      NaN                   NaN                   NaN   
50%                      NaN                   NaN                   NaN   
75%                      NaN                   NaN                   NaN   
max                      NaN                   NaN                   NaN   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
count                     16                    16                    16   
unique                     9                    11                    12   
top       Yoshoku Restaurant                 Diner      Doner Restaurant   
freq                       3                     4                     5   
mean                     NaN                   NaN                   NaN   
std                      NaN                   NaN                   NaN   
min                      NaN                   NaN                   NaN   
25%                      NaN                   NaN                   NaN   
50%                      NaN                   NaN                   NaN   
75%                      NaN                   NaN                   NaN   
max                      NaN                   NaN                   NaN   

       8th Most Common Venue 9th Most Common Venue  \
count                     16                    16   
unique                    12                    11   
top               Donut Shop   Dumpling Restaurant   
freq                       4                     4   
mean                     NaN                   NaN   
std                      NaN                   NaN   
min                      NaN                   NaN   
25%                      NaN                   NaN   
50%                      NaN                   NaN   
75%                      NaN                   NaN   
max                      NaN                   NaN   

             10th Most Common Venue  
count                            16  
unique                           12  
top     Eastern European Restaurant  
freq                              5  
mean                            NaN  
std                             NaN  
min                             NaN  
25%                             NaN  
50%                             NaN  
75%                             NaN  
max                             NaN

#### Cluster 2

In [174]:
london_merged.loc[london_merged['Cluster Labels'] == 1, london_merged.columns[[1] + list(range(5, london_merged.shape[1]))]]

Borough  Cluster Labels          1st Most Common Venue  \
1                      City               1                           Café   
2               Westminster               1                     Restaurant   
3                   Bromley               1                    Pizza Place   
4                 Islington               1                           Café   
6                   Enfield               1                     Steakhouse   
7                Wandsworth               1                           Café   
8                 Southwark               1             Italian Restaurant   
9                      City               1             Italian Restaurant   
10     Richmond upon Thames               1                     Restaurant   
12               Wandsworth               1                           Café   
16                 Lewisham               1             Turkish Restaurant   
20                     City               1                 Sandwich Place   
23            Tower Hamlets               1             Italian Restaurant   
36                   Barnet               1               Sushi Restaurant   
38            Tower Hamlets               1                     Restaurant   
39                   Camden               1  Vegetarian / Vegan Restaurant   
40            Tower Hamlets               1                 Sandwich Place   
41                   Newham               1             English Restaurant   
46              Westminster               1                           Café   
48   Kensington and Chelsea               1            Japanese Restaurant   
49                   Barnet               1             Italian Restaurant   
50              Westminster               1                   Burger Joint   
51           Waltham Forest               1             Italian Restaurant   
54                Islington               1                           Café   
56                   Merton               1              Indian Restaurant   
58              Westminster               1              French Restaurant   
65                  Hackney               1                     Restaurant   
69                 Lewisham               1            Japanese Restaurant   
70                    Brent               1                Thai Restaurant   
71                Southwark               1                     Restaurant   
72                   Ealing               1                           Café   
73   Kensington and Chelsea               1                           Café   
75                Southwark               1                    Pizza Place   
80                Southwark               1                    Pizza Place   
82         Islington & City               1             Italian Restaurant   
84                Islington               1                 Sandwich Place   
86                   Camden               1             Italian Restaurant   
91                   Camden               1                           Café   
103                 Hackney               1                         Bakery   
109                  Ealing               1                           Café   
110                   Brent               1      Middle Eastern Restaurant   
111                Haringey               1             Turkish Restaurant   
117                Lewisham               1                           Café   
118                  Camden               1                 Sandwich Place   
124                 Hackney               1          Vietnamese Restaurant   
125                  Barnet               1           Fast Food Restaurant   
127               Islington               1              French Restaurant   
132               Greenwich               1              Fish & Chips Shop   
134    Camden and Islington               1                           Café   
137             Westminster               1                           Café   
141           Tower Hamlets               

In [181]:
cluster_2 = london_merged.loc[london_merged['Cluster Labels'] == 1, london_merged.columns[[1] + list(range(5, london_merged.shape[1]))]]
cluster_2.describe(include='all')

Borough  Cluster Labels 1st Most Common Venue  \
count             103           103.0                   103   
unique             28             NaN                    26   
top     Tower Hamlets             NaN                  Café   
freq               10             NaN                    29   
mean              NaN             1.0                   NaN   
std               NaN             0.0                   NaN   
min               NaN             1.0                   NaN   
25%               NaN             1.0                   NaN   
50%               NaN             1.0                   NaN   
75%               NaN             1.0                   NaN   
max               NaN             1.0                   NaN   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
count                    103                   103                   103   
unique                    33                    32                    33   
top               Restaurant    Italian Restaurant                  Café   
freq                      12                    13                     9   
mean                     NaN                   NaN                   NaN   
std                      NaN                   NaN                   NaN   
min                      NaN                   NaN                   NaN   
25%                      NaN                   NaN                   NaN   
50%                      NaN                   NaN                   NaN   
75%                      NaN                   NaN                   NaN   
max                      NaN                   NaN                   NaN   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
count                    103                   103                   103   
unique                    44                    47                    40   
top           Sandwich Place      Doner Restaurant           Pizza Place   
freq                       7                     7                     8   
mean                     NaN                   NaN                   NaN   
std                      NaN                   NaN                   NaN   
min                      NaN                   NaN                   NaN   
25%                      NaN                   NaN                   NaN   
50%                      NaN                   NaN                   NaN   
75%                      NaN                   NaN                   NaN   
max                      NaN                   NaN                   NaN   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
count                    103                   103                    103  
unique                    42                    41                     42  
top       Falafel Restaurant  Ethiopian Restaurant    Empanada Restaurant  
freq                       8                     8                     11  
mean                     NaN                   NaN                    NaN  
std                      NaN                   NaN                    NaN  
min                      NaN                   NaN                    NaN  
25%                      NaN                   NaN                    NaN  
50%                      NaN                   NaN                    NaN  
75%                      NaN                   NaN                    NaN  
max                      NaN                   NaN                    NaN

#### Cluster 3

In [175]:
london_merged.loc[london_merged['Cluster Labels'] == 2, london_merged.columns[[1] + list(range(5, london_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0                 Greenwich               2                  Café   
62                  Bromley               2                  Café   
96                   Barnet               2                Bakery   
116                  Camden               2                Bakery   
120               Islington               2                  Café   
138                Lewisham               2                  Café   
139                 Lambeth               2    Italian Restaurant   
162    Richmond upon Thames               2            Restaurant   
176  Kensington and Chelsea               2    Italian Restaurant   
177               Southwark               2           Pizza Place   
178                  Barnet               2                  Café   
194              Wandsworth               2                  Café   
195                   Brent               2                  Café   
203           Tower Hamlets               2  Fast Food Restaurant   
233               Southwark               2                  Café   
234                  Camden               2                  Café   
280                  Barnet               2           Pizza Place   

    2nd Most Common Venue 3rd Most Common Venue       4th Most Common Venue  \
0       Fish & Chips Shop    Yoshoku Restaurant            Currywurst Joint   
62         Breakfast Spot    Yoshoku Restaurant                        Food   
96                   Café    Yoshoku Restaurant                        Food   
116     Indian Restaurant                  Café                   Gastropub   
120  Fast Food Restaurant    Chinese Restaurant                      Bakery   
138         Deli / Bodega     Indian Restaurant                      Bakery   
139        Sandwich Place                  Café                      Bakery   
162             Gastropub                  Café          Yoshoku Restaurant   
176            Restaurant                Bakery               Deli / Bodega   
177                  Café    Chinese Restaurant                      Bakery   
178   Fried Chicken Joint    Yoshoku Restaurant           Fish & Chips Shop   
194    Italian Restaurant            Restaurant              Sandwich Place   
195    Italian Restaurant      Tapas Restaurant  Modern European Restaurant   
203          Burger Joint     Indian Restaurant         Fried Chicken Joint   
233  Fast Food Restaurant           Pizza Place          Chinese Restaurant   
234    Italian Restaurant            Restaurant                      Bakery   
280                  Café    Yoshoku Restaurant            Currywurst Joint   

    5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
0      Dim Sum Restaurant                 Diner       Doner Restaurant   
62                  Diner      Doner Restaurant             Donut Shop   
96                  Diner      Doner Restaurant             Donut Shop   
116    Yoshoku Restaurant    English Restaurant   Fast Food Restaurant   
120      Kebab Restaurant        Sandwich Place   Ethiopian Restaurant   
138    English Restaurant   Filipino Restaurant   Fast Food Restaurant   
139   Japanese Restaurant     Korean Restaurant            Pizza Place   
162   Filipino Restaurant    Dim Sum Restaurant                  Diner   
176           Pizza Place      Tapas Restaurant                   Café   
177    Yoshoku Restaurant    English Restaurant   Fast Food Restaurant   
178    Dim Sum Restaurant                 Diner       Doner Restaurant   
194                Bakery           Pizza Place       Asian Restaurant   
195        Breakfast Spot     French Restaurant                 Bakery   
203           Pizza Place    Italian Restaurant            Snack Place   
233   American Restaurant     Fish & Chips Shop  Vietnamese Restaurant   
234   Japanese Restaurant  Fast Food Restaurant      Indian Restaurant   
280    Dim Sum Restaurant                 Diner       Doner Restaurant   

         8th Most Com

In [182]:
cluster_3 = london_merged.loc[london_merged['Cluster Labels'] == 2, london_merged.columns[[1] + list(range(5, london_merged.shape[1]))]]
cluster_3.describe(include='all')

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
count       17            17.0                    17                    17   
unique      13             NaN                     6                    12   
top     Barnet             NaN                  Café    Italian Restaurant   
freq         3             NaN                     9                     3   
mean       NaN             2.0                   NaN                   NaN   
std        NaN             0.0                   NaN                   NaN   
min        NaN             2.0                   NaN                   NaN   
25%        NaN             2.0                   NaN                   NaN   
50%        NaN             2.0                   NaN                   NaN   
75%        NaN             2.0                   NaN                   NaN   
max        NaN             2.0                   NaN                   NaN   

       3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
count                     17                    17                    17   
unique                     8                    11                    11   
top       Yoshoku Restaurant                Bakery    Dim Sum Restaurant   
freq                       5                     5                     3   
mean                     NaN                   NaN                   NaN   
std                      NaN                   NaN                   NaN   
min                      NaN                   NaN                   NaN   
25%                      NaN                   NaN                   NaN   
50%                      NaN                   NaN                   NaN   
75%                      NaN                   NaN                   NaN   
max                      NaN                   NaN                   NaN   

       6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
count                     17                    17                    17   
unique                    13                    12                    11   
top                    Diner      Doner Restaurant            Donut Shop   
freq                       3                     3                     3   
mean                     NaN                   NaN                   NaN   
std                      NaN                   NaN                   NaN   
min                      NaN                   NaN                   NaN   
25%                      NaN                   NaN                   NaN   
50%                      NaN                   NaN                   NaN   
75%                      NaN                   NaN                   NaN   
max                      NaN                   NaN                   NaN   

       9th Most Common Venue       10th Most Common Venue  
count                     17                           17  
unique                    12                           11  
top     Ethiopian Restaurant  Eastern European Restaurant  
freq                       3                            4  
mean                     NaN                          NaN  
std                      NaN                          NaN  
min                      NaN                          NaN  
25%                      NaN                          NaN  
50%                      NaN                          NaN  
75%                      NaN                          NaN  
max                      NaN                          NaN

#### Cluster 4

In [176]:
london_merged.loc[london_merged['Cluster Labels'] == 3, london_merged.columns[[1] + list(range(5, london_merged.shape[1]))]]

Borough  Cluster Labels  1st Most Common Venue  \
29           Brent               3      Indian Restaurant   
30         Lambeth               3      Indian Restaurant   
47       Greenwich               3      Indian Restaurant   
57          Barnet               3      Indian Restaurant   
63   Tower Hamlets               3      Indian Restaurant   
66        Dartford               3      Indian Restaurant   
83          Barnet               3      Indian Restaurant   
93         Lambeth               3      Indian Restaurant   
97         Enfield               3      Indian Restaurant   
113        Lambeth               3                   Café   
151    Westminster               3                   Café   
152        Hackney               3   Fast Food Restaurant   
155    Westminster               3     Italian Restaurant   
157      Greenwich               3      Indian Restaurant   
159      Greenwich               3      Indian Restaurant   
166          Brent               3  Portuguese Restaurant   
167       Lewisham               3                   Café   
169         Barnet               3      Fish & Chips Shop   
179  Tower Hamlets               3                   Café   
213      Redbridge               3                   Café   
239     Wandsworth               3      Indian Restaurant   
251         Newham               3      Indian Restaurant   
266         Bexley               3      Indian Restaurant   
267         Barnet               3      Indian Restaurant   
274  Tower Hamlets               3      Indian Restaurant   
281      Greenwich               3      Indian Restaurant   

    2nd Most Common Venue     3rd Most Common Venue  \
29           Burger Joint      Fast Food Restaurant   
30              BBQ Joint          Tapas Restaurant   
47        Thai Restaurant                      Café   
57            Pizza Place        Chinese Restaurant   
63                   Café               Pizza Place   
66                   Café        Yoshoku Restaurant   
83                   Café      Fast Food Restaurant   
93     Italian Restaurant            Breakfast Spot   
97                  Diner        English Restaurant   
113     Fish & Chips Shop               Pizza Place   
151       Thai Restaurant             Deli / Bodega   
152                  Café                Restaurant   
155                  Café         French Restaurant   
157    Turkish Restaurant  Mediterranean Restaurant   
159    Italian Restaurant                      Café   
166        Sandwich Place                      Café   
167             Gastropub           Thai Restaurant   
169     Indian Restaurant                      Café   
179    Turkish Restaurant      Fast Food Restaurant   
213    Italian Restaurant        Chinese Restaurant   
239                Bakery                      Café   
251                Bakery            Sandwich Place   
266    Yoshoku Restaurant                      Food   
267      Kebab Restaurant               Pizza Place   
274        Sandwich Place      Pakistani Restaurant   
281                  Café        Yoshoku Restaurant   

           4th Most Common Venue 5th Most Common Venue  \
29            Yoshoku Restaurant    Dim Sum Restaurant   
30                   Pizza Place                  Café   
47          Fast Food Restaurant            Food Truck   
57   Eastern European Restaurant            Food Truck   
63                        Bakery     Fish & Chips Shop   
66             Fish & Chips Shop    Dim Sum Restaurant   
83            Yoshoku Restaurant     Fish & Chips Shop   
93                          Café    English Restaurant   
97            Yoshoku Restaurant                  Food   
113                       Bakery            Restaurant   
151                  Pizza Place    Chinese Restaurant   
152                   Bagel Shop           Pizza Place   
155            Indian Restaurant          Burger Joint   
157         Fast Food Restaurant                  Café   
159    

In [183]:
cluster_4 = london_merged.loc[london_merged['Cluster Labels'] == 3, london_merged.columns[[1] + list(range(5, london_merged.shape[1]))]]
cluster_4.describe(include='all')

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
count       26            26.0                    26                    26   
unique      14             NaN                     6                    15   
top     Barnet             NaN     Indian Restaurant                  Café   
freq         4             NaN                    17                     6   
mean       NaN             3.0                   NaN                   NaN   
std        NaN             0.0                   NaN                   NaN   
min        NaN             3.0                   NaN                   NaN   
25%        NaN             3.0                   NaN                   NaN   
50%        NaN             3.0                   NaN                   NaN   
75%        NaN             3.0                   NaN                   NaN   
max        NaN             3.0                   NaN                   NaN   

       3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
count                     26                    26                    26   
unique                    16                    14                    17   
top                     Café     Fish & Chips Shop    Dim Sum Restaurant   
freq                       5                     4                     5   
mean                     NaN                   NaN                   NaN   
std                      NaN                   NaN                   NaN   
min                      NaN                   NaN                   NaN   
25%                      NaN                   NaN                   NaN   
50%                      NaN                   NaN                   NaN   
75%                      NaN                   NaN                   NaN   
max                      NaN                   NaN                   NaN   

       6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
count                     26                    26                    26   
unique                    15                    14                    17   
top                    Diner      Doner Restaurant            Donut Shop   
freq                       6                     7                     7   
mean                     NaN                   NaN                   NaN   
std                      NaN                   NaN                   NaN   
min                      NaN                   NaN                   NaN   
25%                      NaN                   NaN                   NaN   
50%                      NaN                   NaN                   NaN   
75%                      NaN                   NaN                   NaN   
max                      NaN                   NaN                   NaN   

       9th Most Common Venue       10th Most Common Venue  
count                     26                           26  
unique                    15                           14  
top      Dumpling Restaurant  Eastern European Restaurant  
freq                       7                            8  
mean                     NaN                          NaN  
std                      NaN                          NaN  
min                      NaN                          NaN  
25%                      NaN                          NaN  
50%                      NaN                          NaN  
75%                      NaN                          NaN  
max                      NaN                          NaN

#### Cluster 5

In [177]:
london_merged.loc[london_merged['Cluster Labels'] == 4, london_merged.columns[[1] + list(range(5, london_merged.shape[1]))]]

Borough  Cluster Labels  \
5                                        Islington               4   
11                                       Islington               4   
13                                     Westminster               4   
14                                          Ealing               4   
15                                     Westminster               4   
17                                          Camden               4   
18                                       Southwark               4   
19                                   Tower Hamlets               4   
21                                        Lewisham               4   
24                                          Camden               4   
25                                        Haringey               4   
27                                        Haringey               4   
28                                          Barnet               4   
31                                        Lewisham               4   
33   Kensington and Chelsea,Hammersmith and Fulham               4   
34                                           Brent               4   
35                                          Barnet               4   
37                                       Southwark               4   
42                                       Islington               4   
43                            Richmond upon Thames               4   
44                                        Lewisham               4   
45                                          Camden               4   
52                                           Brent               4   
55                                          Barnet               4   
60                                        Lewisham               4   
61                                        Haringey               4   
64                                          Newham               4   
67                                       Islington               4   
68                                       Southwark               4   
74                                      Wandsworth               4   
76                                          Barnet               4   
77                                          Newham               4   
78                            Richmond upon Thames               4   
79                                         Enfield               4   
81                                       Greenwich               4   
85                             Haringey, Islington               4   
87                                          Newham               4   
88                                        Lewisham               4   
89                                        Haringey               4   
90                                          Barnet               4   
92                          Hammersmith and Fulham               4   
94                                          Barnet               4   
95                                          Camden               4   
98                                       Greenwich               4   
100                                       Hounslow               4   
101                                        Hackney               4   
102                                        Hackney               4   
104                                        Hackney               4   
105                                        Hackney               4   
106                                         Barnet               4   
107                         Hammersmith and Fulham               4   
108                                         Camden               4   
112                                         Barnet               4   
114                                 Waltham Forest               4   
115                                      Islington               4   
119                         Kensington and Chelsea               4   
121                                        Hackney   

In [184]:
cluster_5 = london_merged.loc[london_merged['Cluster Labels'] == 4, london_merged.columns[[1] + list(range(5, london_merged.shape[1]))]]
cluster_5.describe(include='all')

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
count      121           121.0                   121                   121   
unique      29             NaN                    21                    27   
top     Barnet             NaN                  Café                  Café   
freq        10             NaN                    75                    15   
mean       NaN             4.0                   NaN                   NaN   
std        NaN             0.0                   NaN                   NaN   
min        NaN             4.0                   NaN                   NaN   
25%        NaN             4.0                   NaN                   NaN   
50%        NaN             4.0                   NaN                   NaN   
75%        NaN             4.0                   NaN                   NaN   
max        NaN             4.0                   NaN                   NaN   

       3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
count                    121                   121                   121   
unique                    34                    36                    44   
top       Italian Restaurant    Italian Restaurant           Pizza Place   
freq                      13                    12                     9   
mean                     NaN                   NaN                   NaN   
std                      NaN                   NaN                   NaN   
min                      NaN                   NaN                   NaN   
25%                      NaN                   NaN                   NaN   
50%                      NaN                   NaN                   NaN   
75%                      NaN                   NaN                   NaN   
max                      NaN                   NaN                   NaN   

       6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
count                    121                   121                   121   
unique                    51                    46                    45   
top     Fast Food Restaurant  Fast Food Restaurant            Donut Shop   
freq                       9                    10                    10   
mean                     NaN                   NaN                   NaN   
std                      NaN                   NaN                   NaN   
min                      NaN                   NaN                   NaN   
25%                      NaN                   NaN                   NaN   
50%                      NaN                   NaN                   NaN   
75%                      NaN                   NaN                   NaN   
max                      NaN                   NaN                   NaN   

       9th Most Common Venue       10th Most Common Venue  
count                    121                          121  
unique                    48                           49  
top      Dumpling Restaurant  Eastern European Restaurant  
freq                      12                           15  
mean                     NaN                          NaN  
std                      NaN                          NaN  
min                      NaN                          NaN  
25%                      NaN                          NaN  
50%                      NaN                          NaN  
75%                      NaN                          NaN  
max                      NaN                          NaN

### 5.2. Visualizing Clusters

Finally, let's visualize the resulting clusters.

In [186]:
# create map
map_clusters = folium.Map(location=[london_lat, london_lon], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(london_merged['Latitude'], london_merged['Longitude'], london_merged['Location'], london_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster+1), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

**MAP LEGEND**  
Cluster 1 - red dots  
Cluster 2 - purple dots  
Cluster 3 - blue dots  
Cluster 4 - green dots  
Cluster 5 - orange dots  

## 6. Discussion

Analyzing the most popular restaurants in each cluster, the stakeholder should prefer the *least* popular types as a safe choice. There is no sense in opening the 17th pizzeria in the same street. Of course, there might be more than 10 types in a location. And one might object, that following this logic, the stakeholder must prefer the last type in a full list, and not the 10th one. But bear in mind that descending on the popularity list we might face an absence of demand for this type of food, and open a restaurant that is not needed in this particular location. Presence of interested customers is a must for a successful business. That is why in our recommendations we offer to stop on 10th and 9th positions.

Recommendations, based on description of each cluster:  
**Cluster 1 Locations:** Eastern European or Dumpling Restaurant  
**Cluster 2 Locations:** Empanada or Ethiopian Restaurant  
**Cluster 3 Locations:** Eastern European or Ethiopian Restaurant  
**Cluster 4 Locations:** Eastern European or Dumpling Restaurant  
**Cluster 5 Locations:** Eastern European or Dumpling Restaurant  

After the type of restaurant is chosen, it is time to select a right place. Using the map created in 5.2 and its legend the solution is quite obvious. 

## 7. Conclusion

In this report we worked out a methodology to determine what the most promising type of restaurant is and where it should be opened.  

We collected information about London boroughs from Wikipedia, and using geospatial libraries mapped them. Using Foursquare API, we collected the top 100 restaurants and their types for each location within a radius 500 meters from its central point. Then we grouped collected restaurants by location and by taking the mean of the frequency of occurrence of each type, preparing them for clustering. Finally we clustered restaurants by the k-means algorithm and analize the top 10 most common restaurants in each cluster, making useful observations. Eventually we visualized clusters on the map, thus showing the best locations for opening the chosen type of restaurant.

This type of analysis can be applied to any city of your choice that has available geospatial information.

This type of analysis can be applied to any type of venue (shopping, clubs, etc.) that is available in Foursquare database.